src/Polly.Strategies/Retry/RetryStrategy.cs

```diff
@@ -0,0 +1,96 @@
+using Polly.Internals;
+
+namespace Polly.Retry;
+
+internal sealed class RetryStrategy : DelegatingResilienceStrategy
+{
+    private readonly RetryStrategyOptions _options;
+    private readonly PredicatesHandler<ShouldRetryArguments> _shouldRetry;
+    private readonly EventsHandler<RetryActionArguments> _onRetry;
+    private readonly int _maxRetries;
+
+    public RetryStrategy(RetryStrategyOptions options)
+    {
+        _options = options;
+        _shouldRetry = PredicatesHandler<ShouldRetryArguments>.Create(options.ShouldRetry);
+        _onRetry = EventsHandler<RetryActionArguments>.Create(options.OnRetry);
+
+        _maxRetries = options.RetryCount;
+    }
+
+    public async override ValueTask<T> ExecuteAsync<T, TState>(Func<ResilienceContext, TState, ValueTask<T>> execution, ResilienceContext context, TState state)
+    {
+        var tryCount = 0;
+
+        while (true)
+        {
+            context.CancellationToken.ThrowIfCancellationRequested();
+
+            bool canRetry;
+            Outcome<T> outcome;
+
+            try
+            {
+                var result = await base.ExecuteAsync(execution, context, state).ConfigureAwait(context.ContinueOnCapturedContext);
+                outcome = new Outcome<T>(result);
+
+                if (!await _shouldRetry.HandleAsync(outcome, new ShouldRetryArguments(tryCount, context), context).ConfigureAwait(context.ContinueOnCapturedContext))
+                {
+                    return result;
+                }
+
+                canRetry = tryCount < _maxRetries;
+                if (!canRetry)
+                {
+                    return result;
+                }
+
+                return result;
```

Comment:
Is this code right? Seems to return `result` no matter what `canRetry`'s value is.